### DISCUSSION TITLE
Is it necessary to export a global variable in module federation?

### DISCUSSION THREAD

**Discussion by joe223** [webpack/webpack#14644] [Q&A]
We export an external module called `mfeBBB` in module federation example.

https://github.com/webpack/webpack/blob/a61e83c75b1ac4df4ceaa8dad1fdee94bbb041d9/examples/module-federation/webpack.config.js#L45-L111

And we get the built product as below

```js
var mfeBBB;
/******/ (() => { // webpackBootstrap
/******/ 	"use strict";
/******/ 	var __webpack_modules__ = ([
/* 0 */
/*!***********************!*\
  !*** container entry ***!
  \***********************/
/*! unknown exports (runtime-defined) */
/*! runtime requirements: __webpack_require__.d, __webpack_require__.o, __webpack_exports__, __webpack_require__.e, __webpack_require__, __webpack_require__.* */
/***/ ((__unused_webpack_module, exports, __webpack_require__) => {

var moduleMap = {
...
```

**Comment by alexander-akait** [comment]
Yes, expected, what is the problem?

**Reply by joe223** [reply]
I'm curious about the reason. Global variables may have potential conflicts.

**Reply by alexander-akait** [reply]
For communication purposes, use unique names (webpack do it by default based on `package.json` name)

**Reply by joe223** [reply]
Maintaining these global variable names will increase the complexity. That's why I asked this question.

**Reply by joe223** [reply]
> For communication purposes

Should we add a prefix by default? Such as `__webpack__mf__[package name]`?

**Comment by farhadnowzari** [comment]
I have also somehow the same question.

on both UIs I have global axios.
now when I use a remote component it uses the host axios and not it's own.

is there a solution to this?

**Comment by jayjnu** [comment]
@joe223 

I encountered the same problem and found a workaround.
I would use below method  if module federation plugin is not planned to  support scoped container name.

### Implementing scoped container entry

here is a original federated container entry build result.
**remoteEntry.js**
```js
var my_container;
(() => {
...
  my_container = __webpack_exports__;
});
```

Because the result is just a side effect code that,

1. declares a global variable
2. runs a function (webpack runtime)
3. assigns loaded module to the variable name

It seems possible to wrap the result code like this

```js
(function($scope) {
  var my_container;
  (() => {
  ...
    my_container  = __webpack_exports__;
  });
  $scope.my_container = my_container
})(window['myFederationScope'] = window['myFederationScope'] || {})

```

### Let's test


the host container is [dynamic remote container](https://webpack.js.org/concepts/module-federation/#dynamic-remote-containers).  so i can manipulate how container is resolved in runtime like this interface

**host/loadContainer.js**
```javascript
async function loadContainer(containerName, url) {
  await loadJs(url);
  // resolver knows where the container is scoped
  const container = window['myFederationScope'][containerName];
  return container;
}
```

**init.js**
```js
(async () => {
 const container = await loadContainer('my_container', 'myservice/my_container.main.js');
  // Initializes the shared scope. Fills it with known provided modules from this build and all remotes
  await __webpack_init_sharing__('default');
  const container = window.someContainer; // or get the container somewhere else
  // Initialize the container, it may provide shared modules
  await container.init(__webpack_share_scopes__.default);
  const module = await container.get('./module');
})()
```

Test works as expected

### ScopedFederationPlugin 
To do this automatically, one can implement a dedicated webpack plugin like this

```js
const containerName = 'my_container';
module.exports = {
  plugins: [
    new ModuleFederationPlugin({
     name: containerName,
     filename: 'main.js'
    }),
    new ScopedFederationPlugin({
      name: containerName,
      filename: 'main.js',
      scope: 'myFederationScope'
    })
  ]
}
```
